raiden_libs/network/matrix/client.py

```diff
@@ -0,0 +1,165 @@
+import logging
+from typing import List
+from urllib.parse import quote
+
+import gevent
+from matrix_client.client import CACHE, MatrixClient
+from matrix_client.errors import MatrixRequestError
+from matrix_client.user import User
+
+from .room import Room
+from .utils import geventify_callback
+
+
+logger = logging.getLogger(__name__)
+
+
+class GMatrixClient(MatrixClient):
+    """ Gevent-compliant MatrixClient subclass """
+
+    def __init__(
+            self,
+            base_url,
+            token=None,
+            user_id=None,
+            valid_cert_check=True,
+            sync_filter_limit=20,
+            cache_level=CACHE.ALL
+    ):
+        super().__init__(
+            base_url,
+            token,
+            user_id,
+            valid_cert_check,
+            sync_filter_limit,
+            cache_level
+        )
+        self.should_listen = False
+        self.sync_thread = None
+
+    def listen_forever(self, timeout_ms=30000, exception_handler=None, bad_sync_timeout=5):
+        """
+        Keep listening for events forever.
+        Args:
+            timeout_ms (int): How long to poll the Home Server for before
+               retrying.
+            exception_handler (func(exception)): Optional exception handler
+               function which can be used to handle exceptions in the caller
+               thread.
+            bad_sync_timeout (int): Base time to wait after an error before
+                retrying. Will be increased according to exponential backoff.
+        """
+        _bad_sync_timeout = bad_sync_timeout
+        self.should_listen = True
+        while self.should_listen:
+            try:
+                self._sync(timeout_ms)
+                _bad_sync_timeout = bad_sync_timeout
+            except MatrixRequestError as e:
+                logger.warning('A MatrixRequestError occured during sync.')
+                if e.code >= 500:
+                    logger.warning(
+                        'Problem occured serverside. Waiting %i seconds',
+                        _bad_sync_timeout
+                    )
+                    gevent.sleep(_bad_sync_timeout)
+                    _bad_sync_timeout = min(_bad_sync_timeout * 2, self.bad_sync_timeout_limit)
+                else:
+                    raise
+            except Exception as e:
+                logger.exception('Exception thrown during sync')
```

Comment:
SHould we add some info about the exception that occurred?

Reply:
Definitely, the whole error handling here isn't that great...

Reply:
`logger.exception` logs the exception automatically: https://docs.python.org/3/library/logging.html#logging.Logger.exception

Reply:
Sure, but the whole "catch everything, log, and then dispatch with a function" scheme doesn't feel right. But I guess that's something for later.